### PR TITLE
Add zCash Support

### PR DESCRIPTION
**Pull request by ososco**
Added support for zCash (ZEC) transparent address (t1)

### CODE REVIEW

**Comment by makoto** [comment]
The test seems failing https://travis-ci.com/ensdomains/address-encoder/jobs/257762485#L1167

**Comment by ososco** [comment]
Hi @makoto, i'm working on it right now, i'm not sure what is the issue tho, these are the P2PKH and P2SH addresses provided by Zcash Protocol Specification.

Capture

File: src/index.ts [hunk 1]
```diff
@@ -39,6 +39,45 @@ function makeBase58CheckEncoder(p2pkhVersion: number, p2shVersion: number): (dat
   };
 }
 
+function decodeZecAddr(p2pkhVersions: number[], p2shVersions: number[]): (data: string) => Buffer {
```
Comment (Arachnid) [inline src/index.ts]: How do these differ from `makeBase58CheckEncoder` and `makeBase58CheckDecoder`?

**Comment by ososco** [comment]
Hi @Arachnid,

Zcash has two main types of addresses: transparent (t-addresses which start with “t”), and private shielded (z-addresses which start with “z”).

The code I've submitted is only meant for public transparent addresses (t1 for P2PKH and t3 for P2SH), while the raw encoding of a shielded address is the following:

Capture

To be honest, adding support for zCash shielded addresses is beyond my knowledge unfortunately.

both functions in the code above are a modified copy of _makeBase58CheckEncoder_ and _makeBase58CheckDecoder_, since two bytes are used in zCash for the version field identifying the address type.

Please feel free to close this PR after reading my reply.

**Comment by Arachnid** [comment]
> To be honest, adding support for zCash shielded addresses is beyond my knowledge unfortunately.

I'm afraid the only way we could accept a PR to add support for ZCash transparent addresses only is if it allows a clear backwards-compatible path to supporting shielded addresses too. I don't think the current system does that.

However, this could be as simple as standardising our own 'version byte' before the address that indicates if it's a transparent or shielded address. Since ZCash doesn't appear to have a single canonical binary format for all address types, we will have to just invent one ourselves.

> both functions in the code above are a modified copy of makeBase58CheckEncoder and makeBase58CheckDecoder, since two bytes are used in zCash for the version field identifying the address type.

If you want to pursue this further, could you generalise `makeBase58Check*` to accept multiple bytes, instead of copying the code?

**Comment by ososco** [comment]
I managed to rewrite `makeBase58Check` to make it accept 2 bytes, but i have a question:

```
  bitcoinChain('LTC', 2, 'ltc', [0x30], [0x32, 0x05]),
  ...
  ...
  bitcoinChain('MONA', 22, 'mona', [0x32], [0x37, 0x05]),
```

For both Litecoin and Mona, there is a second byte added as a second p2sh version, however in the function `makeBitcoinEncoder` [only the first key is passed](https://github.com/ensdomains/address-encoder/blob/39a1772587854b98bb8dd326c5c82c2626985445/src/index.ts#L124):

**Comment by ososco** [comment]
I managed to rewrite makeBase58Check to make it accept 2 bytes, but i have a question:

```
  bitcoinChain('LTC', 2, 'ltc', [0x30], [0x32, 0x05]),
  ...
  ...
  bitcoinChain('MONA', 22, 'mona', [0x32], [0x37, 0x05]),
```

For both Litecoin and Mona, there is a second byte added as a second p2sh version [0x05], however in the function makeBitcoinEncoder [only the first key is passed](https://github.com/ensdomains/address-encoder/blob/39a1772587854b98bb8dd326c5c82c2626985445/src/index.ts#L124), which pass it to `makeBase58CheckEncoder*` as a single key, and this second key is not used at all, so a deprecated Litecoin 3-address (3MSvaVbVFFLML86rt5eqgA9SvW23upaXdY) which use P2SH(0x05) is not actually supported.

What is the purpose of this second key in the current code?

**Comment by Arachnid** [comment]
> For both Litecoin and Mona, there is a second byte added as a second p2sh version [0x05], however in function makeBitcoinEncoder only the first key is passed, which pass it to makeBase58CheckEncoder* as a single key, and this second key is not used at all, so a deprecated Litecoin 3-address (3MSvaVbVFFLML86rt5eqgA9SvW23upaXdY) which use P2SH(0x05) is not actually supported.
> 
> What is the purpose of this second key in the current code?

It means that we can decode all forms of address, but addresses are always encoded as the first value, which is treated as the canonical format.

File: src/index.ts [hunk 2]
```diff
@@ -25,13 +25,16 @@ function makeBase58CheckEncoder(p2pkhVersion: number, p2shVersion: number): (dat
         ) {
           throw Error('Unrecognised address format');
         }
-        addr = Buffer.concat([Buffer.from([p2pkhVersion]), data.slice(3, 3 + data.readUInt8(2))]);
+        addr = Buffer.concat([Buffer.from(p2pkhVersion), data.slice(3, 3 + data.readUInt8(2))]);
         return bs58check.encode(addr);
       case 0xa9: // P2SH: OP_HASH160 <scriptHash> OP_EQUAL
         if (data.readUInt8(data.length - 1) !== 0x87) {
           throw Error('Unrecognised address format');
+        } else if (p2pkhVersion.length === 2) {
```
Comment (Arachnid) [inline src/index.ts]: What is the purpose of this? Why check for the length of the p2pkhVersion bytes inside the p2sh encoding branch? Can't you just use the unmodified p2shversion string in each case?

File: src/index.ts [hunk 3]
```diff
@@ -42,10 +45,16 @@ function makeBase58CheckEncoder(p2pkhVersion: number, p2shVersion: number): (dat
 function makeBase58CheckDecoder(p2pkhVersions: number[], p2shVersions: number[]): (data: string) => Buffer {
   return (data: string) => {
     const addr = bs58check.decode(data);
-    const version = addr.readUInt8(0);
-    if (p2pkhVersions.includes(version)) {
+    const version = addr.length === 21 ? addr.slice(0, 1) : addr.slice(0, 2);
```
Comment (Arachnid) [inline src/index.ts]: I think you need to modify this to take a list of versions, each of which is either a buffer or an array of bytes, and check for the version inside that array.

Also, is there some general-purpose way to determine how many version bytes there are, other than just checking the  string length?

**Comment by Arachnid** [comment]
Closing for inactivity.

**Pull request closed without merging**